### ISSUE TITLE
please document, whether generators may create real unit instances

### ISSUE THREAD

**Issue by calestyo** [systemd/systemd#7257]
### Submission type
 - Request for enhancement (RFE)

### systemd version the issue has been seen with
> 235

### Used distribution

Debian

On https://github.com/dCache/dcache/issues/3630 we stumbled across a question which doesn't seem to be clarified in the systemd manpages.

Can one create e.g. within a generator real `foo@instance.service` files (e.g. dynamically as *regular files*) instead of having a `foo@.service` (regular) file and symlinks to it which e.g. specify the instance name.

On IRC I was told this is not supported and this seems to be backed by the fact that one cannot enable such `foo@bar.service` if there is no `foo@.service`.

systemd.generator(7), says e.g.
> Generators
>        can dynamically generate unit files or create symbolic links to unit
>        files to add additional dependencies

It would be nice, if it could be clarified here (or at some proper place), whether it is possible to create (dynamically or not), real (regular) `foo@someInstance.service` files, whether or not there is a corresponding `foo@.service`.

Thanks,
Chris.

**Comment by poettering** [comment]
This is actually supported just fine. It's entirely OK to override only one specific instance of a unit file template, instead of the whole template. And yeah, I guess we could document this.

**Comment by calestyo** [comment]
Hey Lennart.

I see... I assume it's then also okay if one has in addition to the real regular file `foo@instance.service` the template `foo@.service`, right?

And as a side question, what's then the correct solution (from a systemd POV) to what I've been discussion in some other project here https://github.com/dCache/dcache/issues/3630 about what they're doing at their generated instance service files from https://github.com/dCache/dcache/blob/master/skel/lib/systemd/system-generators/dcache-generator
What they do is basically: have no `foo@.service` at all but only regular files `foo@instance.service`.

This partially works fine, but when I try to do
`# systemctl enable foo@instance.service`
it complains:
`Failed to enable unit: Unit /run/systemd/generator/dcache@admin.service is transient or generated.`

Is this then just a "bug" in systemd (if the above is supported as you say), or would the documentation need to state something like:
You can override certain instance untis like `foo@instance.service`, but you still need to have a template instance unit `foo@.service`, i.e. when using instance units (unti files with `@`) there always need to be a template instance unit `foo@.service`.

Thanks,
Chris.

**Comment by boucman** [comment]
naive question, but what would be the point of having an template instace without a template... afaict a template instace with no template is just a normal unit, right ?

**Comment by calestyo** [comment]
Well what dcache (a storage manamgent system for storage in the PiB range) is doing is basically the following:
It can run arbitrary dcache services (called domains) on each node in a cluster. Which are run and how they're configured is determined by dCache's configuration.

So one will get e.g.
```
dcache@ftp.service
dcache@pool.service
dcache@nfs.service
```
each being one instance of dCache, doing different things.

The reason that the original author had to not simply make one `dcache@.service` file, using `%i` and `%I` and the like, was that dCache's config already has settings for which user to run at, or in which interval to restart after the process died.
(This in turn simply came from the fact, that in sys-v-init days there was no defined way like systemd's `User=`, so they simply put it in their own config, instead of something like Debian's `/etc/default/*`.)

Now when they made the first iteration of systemd support, they rather tried to keep these dCache config settings (there is e.g. a `dcache.user` in dCache's config).
But as one cannot write something like
User=$( someHowGetTheValueOf_dcacheUser ), they made a `dcache@domain.service` file for each configured domain from within the generator, without having a `dcache@.service` file at all (and without the usual symlinking approach).


So conceptually, I think, the domains of dCache are just what the instances of systemd are, as always the same code is started, just with some additional config, which decide what will actually be loaded by the Java.
But from the systemd POV they were not really using the @-templates.

And that's the story how we at dCache fell in that trap.

(I already proposed them, that settings like dcache.user should be actually pulled out of dCache and canonically placed in systemd.. i.e. dCache admins would no longer set `dcache.user,` but override `User=` but getting that done will take a while ;) ).

**Comment by calestyo** [comment]
btw:
There is this mechanism of just overriding selective parts of a unit file, as described in:
systemd.unit(5) manpage... section "Overriding vendor settings", where one would create e.g. a:
`/etc/systemd/system/dcache@someInstance.service.d/local.conf`
to just override e.g. `User=` instead of the whole unit file.

It's not excluded in the documentation... but is this also supported for dynamically created instance units?

**Comment by boucman** [comment]
I still don't understand... the point of templates is to dynamically create similar units, the list is not known at systemd startup time.

if the units are known at startup time, then you can use generators to create normal units namde dcache-ftp.service dcache-pool.service etc.

You could probably create an empty dcache@.service to help systemd deal correctly with what you need to do (not tested) but I don't really see the point of using the @ syntax if there is no common code to put in the template

It is perfectly fine for a generator to read dcache's configuration files and generate units according to that config file, wouldn't that do the trick for you ?

there is probably something I don't understand of your use-case, I never had a need for anything like dcache myself, so I could miss something obvious

**Comment by calestyo** [comment]
Hmm... well using generated `dcache-someDomain.service` files would work, I guess.
The `@` thingy just felt to be like a conceptual match, as it's about instances of the same thing.
Similar, e.g. to what Debian's postgresql packages to... they also have a postgresql.service and a postgresql@.service, with sylinks like postgresql@version-clustername.service being created by their generator... but not really dynamically, but rather only when daemon-reload is called (which I guess their pg_* cluster management scripts do as well).

**Comment by boucman** [comment]
well, what postgresql does is a bit complicated but makes sense

* postgresql.service and postgresql@.service are "totally unrelated" the happen to have the same name and to both come from postgresql, but systemd doesn't look for that, so you have a normal instance and a template
* creating a service (or a symlink in this case) named after an instance of a template allows to override that particular instance without influencing other instances (which will be based purely on the template)
* systemd automatically calls generators to create new units at reaload time, it's the right point to do that

So yes, @ is really a magic character for systemd and from what you describe, I think that generated units with "normal" names are a better match for what you are trying to do. Templates are only usefull for creating on-demand, at runtime (not reload-time) units, and that does not seem to be what you are trying to do...

**Comment by calestyo** [comment]
Well AFAIU, postgresql.service *is* related to postgresql@.service in that the former pulls in all the later, so when I say systemctl start postgresql.service, all instances of postgresql@.serice get started.

Mhh... but then postgresql would also kind mis-use this, right?
Apart from that... from the dCache-side, there's the following: a cached version of the config in some location, which is generated (at certain times) and actually used instead of the stuff in /etc/dcache/*
So I think dCache would be happy if any changes to that (e.g. new domains) would be picked up by systemd dynamically (without having to call daemon-reload).

**Comment by boucman** [comment]
if it's really dynamic, only decided at runtime, and all instances are completely different... I guess your best approch might be to create the units dynamically with systemd's dbus API... but i don't know it well enough to help you there...

Anyway, this is a bug tracker, the original (documentation) bug makes sense, but you should probably wirte to the ML if you want to ask a support-type question...

**Comment by fsateler** [comment]
@calestyo 

>Is this then just a "bug" in systemd (if the above is supported as you say), or would the documentation need to state something like:
> You can override certain instance untis like `foo@instance.service`, but you still need to have a template instance unit `foo@.service`, i.e. when using instance units (unti files with @) there always need to be a template instance unit `foo@.service`.

I think there's a misunderstanding about what the error message is telling you:

```
Failed to enable unit: Unit /run/systemd/generator/dcache@admin.service is transient or generated.
```

The error is because systemctl is warning you that you are enabling for the next boot a unit that may not exist, as it is runtime only. One could argue about whether the error is warranted or not, but the error is in no way unrelated to the missing `dchache@.service`. The reason th error disappeared when you created it is because you created it in `/etc`, and thus systemd won't complain anymore (now there is a unit that will exist on next boot). You can test this assertion by creating `dcache@.service` in `/run/systemd` instead of `/etc/systemd`, and see that the error persists.

As a side note, I'd recommend changing dcache to have a persistent, static `dcache@.service` with the default options, and have the generator simply generate overrides for the relevant knobs.

**Comment by calestyo** [comment]
> but the error is in no way unrelated to the missing dchache@.service. The reason th error disappeared when you created it is because you created it in /etc, and thus systemd won't complain anymore (now there is a unit that will exist on next boot).

Uhm,... so AFAIU, the existence of foo@.service in a "permanent" location (e.g. /etc or /lib) makes any transient foo@instance.service (whether regular fiel or symlink) in /run okay to be enabled, right?

But that still means, that I do need a foo@.service in some (permanent) location… I'd guess this is because systemd thinks (and is in principle right with it), that even if the transient foo@instance.service would be gone after reboot, it can use the foo@.service to re-create the service (which of course may then lack any of the possibly necessary overriding customisations that a foo@instance.service made by a generator may have had in /run).

Does that sound reasonable?

If so, then for improving the documentation one could write about the following:
- systemd allows generators to create foo@instance.service files, which (I assume) count conceptually as overriding the foo@.service, even if the foo@.service does not really exist
- if one wants to systemctl enable such customising foo@instance.service, that one itself needs to be either permanent (i.e. in /etc/ or /lib) _OR_ there need to be permanent foo@.service.
- one can even override these foo@instance.service in one level further down, by creating e.g.  /etc/systemd/system/foo@instance.service.d/local.conf.

And one should possibly further note, that _if_ the foo@instance.service is created by a generator as regular file (with customisations), that these are of course possibly lost after reboot and recreating them from the foo@.service template may not be possible… and that therefore users are strongly advised to rather create proper template foo@.service and just symlink to that using %i and friends in the template.

Right?


And as for the specific case of dCache.
My personal solution would have been this:
- pull out the options like dcache.user from dCache and move them in the place where they should be canonically set (namely some systemd unit)… and if admins of dCache want to set e.g. the system user a dCache domain runs under, let them create /etc/systemd/system/dcache@someInstance.service.d/local.conf where they e.g. override User=
- create a proper dcache@.service template unit
- let the dcache generator create symlinks to that for each instance
- let the dcache generator create an EnvironmentFile in some temp location for each domain, where settings like CLASSPATH and so on are stored with the per-domain values (pulled out from dCache's config)

What you propose "and have the generator simply generate overrides for the relevant knobs."… seems to be an approach, that allows to leave configs like dcache.user within dCache.
But I think that should anyway be split out of it.
Also it would probably require that one can make something like:  /etc/systemd/system/dcache@someInstance.service.d/local.conf in /run … does that work?

Thanks,
Chris.

**Comment by poettering** [comment]
I think it's conceptually a bad idea to use template/instances if you don't want to actually dynamically instantiate multiple very similar instances easily. If you generate stuff, and never want to define a template, then it's really the wrong tool to use. That said, for instances to work, there actually doesn't have to be a template file around, it's entirely sufficient if for every instance you reference you have a fully defined instance unit file around.

But again: unless you actually want templating don't use "@" in your unit names. It's just confusing to users, and semantically skewed. It suggests users can do "systemctl enable yourservice@foobar.service" to get another instance of something you defined, but in fact your are really misleading them there...

So, I think your usecase is flawed, but then again, technically systemd does allow such use, and we should probably document that in the generator docs

**Comment by poettering** [comment]
Documentation fix waiting in #7342
